### DISCUSSION TITLE
✨ Improved Cache visibility

### DISCUSSION THREAD

**Discussion by c-p-b** [dagger/dagger#12498] [Ideas]
### What are you trying to do?

Part of the story around Dagger's appeal is its cache-first approach, backed by Buildkit.

Dagger Cloud, for those who have alpha access, offers some visibility features, namely a UI that shows whether something was cached or not. However, at Airbyte, where we leverage Dagger heavily, we have noticed limitations around the existing tooling that Dagger offers. Namely, we are missing some features that allow us to automatically analyze and qualify builds, as well as some UI/UX around how to access that data.

 
### Why is this important to you?

Cache visibility is critical for qualification of builds and construction of efficient builds. For many, Dagger's caching mechanisms are one of the core draws that attract them to the platform.

End users such as ourselves need to be sure that when we ship code that we are shipping up-to-date code and not something that was accidentally cached.  We also need to be sure that we are efficiently caching as much as possible. In other words, we need to make sure that we are not overcaching or undercaching. What follows is a non-exhaustive list of feature request ideas that we have discussed to better give some developers tools to handle this, that perhaps some UX could be built on top of. Other feature request suggestions are welcome, ideally this should start as a discussion point rather than be interpreted as a list of exact feature asks - just ideas to put people in the headspace of the problems we are trying to solve.

- Analyze OCI-compatible images built with Dagger to determine whether the layers derived in them were from cache or were built
- Provide a way to store and retrieve at build time whether buildkit layers were retrieved from cache or not
- Expose cache related metadata on File or Directory objects
- If it is possible, deterministically expose what "busted" a given layer cache (even if doing so is very slow and sometimes useless it could be hidden behind some debug flag) 
- Provide general guidelines/documentation for preventing over-caching
- Provide general guidelines/documentation for preventing under-caching

### How are you currently working around this?

Currently we have to deeply investigate cache issues, and it's also tough for developers newer to Dagger to understand what is getting cached and why, leading people to write sometimes inefficient implementations that could easily be improved with correct knowledge. In particular knowing what is/isn't busting the cache is something that is not very easily exposed in Dagger's documentation or UX.

**Comment by d3rp3tt3** [comment]
Thank you for your feedback, @cpdeethree! 

It's on our roadmap for the UI to

*   Surface caching analytics across changes
*   Highlight caching analytics in a change
*   Make it easy to determine why something wasn't cached, debug it, fix it, and then verify it's now cached

For determining why something wasn't cached, there are two places in the UI we've considered adding more visibility: the tree view and the DAG view. You can see relations in the DAG view, but it's experimental and needs some improvements for usability. In the near term, we'd like to add more information to each step in the tree view, such as the inputs to each node. We'll soon do more comprehensive UX prototyping for cache status and metadata in the UI. I'd like to pass a current UX mock-up by you for showing node inputs and also set aside time to talk specifically about cache visibility.

**Comment by c-p-b** [comment]
> Thank you for your feedback, @cpdeethree! 
> 
> It's on our roadmap for the UI to
> 
> * Surface caching analytics across changes
> * Highlight caching analytics in a change
> * Make it easy to determine why something wasn't cached, debug it, fix it, and then verify it's now cached
> 
> For determining why something wasn't cached, there are two places in the UI we've considered adding more visibility: the tree view and the DAG view. You can see relations in the DAG view, but it's experimental and needs some improvements for usability. In the near term, we'd like to add more information to each step in the tree view, such as the inputs to each node. We'll soon do more comprehensive UX prototyping for cache status and metadata in the UI. I'd like to pass a current UX mock-up by you for showing node inputs and also set aside time to talk specifically about cache visibility.

Awesome! Are there any plans to expose some extra caching information in the CLI/API as well? We would like to have some automation around cache reporting and analysis as well

**Comment by d3rp3tt3** [comment]
> Awesome! Are there any plans to expose some extra caching information in the CLI/API as well? We would like to have some automation around cache reporting and analysis as well

I don't currently have it recorded as a request. How do you envision using the telemetry? Would it be in a report from CI?

**Comment by c-p-b** [comment]
In a pairing session with @vito yesterday we figured one easy low hanging fruit for improvement in regards to undercaching was to provide a builtin way to view the top X files most recently modified in a container. Currently you can manually do that with a command like this:


(Edited for better performance find command)
```bash
exec /bin/sh -c find . -type f -printf '%T@ %p\n' | sort -n -k 1,1 -r | awk '{print strftime("%Y-%m-%d %H:%M:%S", $1), $2}' | head -n 10
```
We were able to successfuly use that in my case to determine some gradle lock files were busting the cache during my runs.

But of course it's annoying to have to do that in containers manually, especially if you have some nested layering and have to walk down the container stack 4 or 5 deep to get to where that function needs to go.

We figured that this is one of the most common ways that cache gets busted and just having a builtin way to visualize that along with a guideline on how to use it would go a long way towards helping find out where things are being undercached.

**Comment by jpadams** [comment]
related https://github.com/dagger/dagger/issues/2164
